### PR TITLE
Branching I/O Manager

### PR DESCRIPTION
**Pull request by schrockn**
### Summary & Motivation

Stemming from a conversation with Daniel Gafni.

A common problem in ML pipelines (and data pipelines in general) is that you want to develop on upstream production assets but write to a dev or branch environment during iterative development. Sometimes the source data for an ML model is the result of expensive and long-running computations and rerunning those is very expensive. You want to be able to use those production assets and then fork storage into a staging environment for the assets one wants to iterate on.

Data warehouses like Snowflake support this natively with copy-on-write CLONE features. However in the generalized case one needs to control this with software. This is what this PR seeks to enable.

It works by leveraging Dagster's metadata to "mark" certain materializations as created in dev and then creating a BranchingIOManager which composes two i/o managers, one for the branch environment and one for the parent environment. The handle_output method of BranchingIOManager writes to the branch environment unconditionally, and creates an AssetMaterialization with an "io_manager_branch" entry (is configurable) that marks it as "dev" (again, configurable). The `load_input` checks for a previous AssetMaterialization marked in that branch. If that exists, it reads from the branch. Else it reads from parent.

This also leverages the new Definitions API to make this more tractable, as it relies on object references to created IOManagers, rather than @io_manager-decorated functions. We can simply pass in created resources rather than relying on the `required_resource_key` system.

### How I Tested These Changes

Included tests

### CODE REVIEW

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11315** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11315" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11315?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/execution/context/input.py [hunk 1]
```diff
@@ -149,6 +152,14 @@ def __del__(self):
         if self._resources_cm and self._resources_contain_cm and not self._cm_scope_entered:
             self._resources_cm.__exit__(None, None, None)  # pylint: disable=no-member
 
+    @property
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/context/input.py:155]: This will have to be done more thoughtfully in a separate PR. Not having the instance available to InputContext is an oversight IMO.

File: python_modules/dagster/dagster_tests/storage_tests/test_staging_io_manager.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+        assert get_text_metadata(asset_mat.metadata_entries, label="another_env") == "staging"
+
+
+@asset
+def now_time_plus_20_after_plus_N(now_time_plus_N: int) -> int:
+    return now_time_plus_N + 20
+
+
+def test_basic_workflow():
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/storage_tests/test_staging_io_manager.py:155]: This is the test that demonstrates the workflow where one is iterating on asset in the middle of an asset graph.

File: python_modules/dagster/dagster/_core/execution/context/input.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/execution/context/input.py:155]: Yeah - filed this recently: https://github.com/dagster-io/dagster/issues/11016

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        return None
+
+    return (
+        dagster_event.event_specific_data.asset_observation
+        if isinstance(dagster_event.event_specific_data.asset_observation, AssetObservation)
+        else None
+    )
+
+
+def get_text_metadata(
```
Comment (sryza) [inline python_modules/dagster/dagster/_experimental/staging_io_manager.py:53]: `AssetMaterialization` now has a `metadata` property that I believe should make this more straightforward

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            entry_to_search.entry_data, TextMetadataValue
+        ):
+            return entry_to_search.entry_data.value
+
+    return None
+
+
+class DualEnvironmentIOManager(IOManager):
+    def __init__(
+        self, *, prod_io_manager, staging_io_manager, env_key="env", staging_label="staging"
```
Comment (sryza) [inline python_modules/dagster/dagster/_experimental/staging_io_manager.py:70]: "env" is a very general term. Would bias towards trying to find something more specific - e.g. "storage_env".

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    MetadataEntry,
+    PartitionMetadataEntry,
+    TextMetadataValue,
+)
+from dagster._core.events import AssetObservationData, DagsterEventType, StepMaterializationData
+from dagster._core.events.log import EventLogEntry
+from dagster._core.storage.io_manager import IOManager
+
+
+def get_asset_materialization_from_event_log_entry(
```
Comment (sryza) [inline python_modules/dagster/dagster/_experimental/staging_io_manager.py:15]: IMO we should make this a property on `EventLogEntry`. We've already done so with `EventLogRecord`.

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                    asset_materialization.metadata_entries, label=self.staging_label
+                )
+                if env_entry:
+                    return self.staging_io_manager.load_input(context)
+
+        return self.prod_io_manager.load_input(context)
+
+    def handle_output(self, context: OutputContext, obj: Any) -> None:
+        self.staging_io_manager.handle_output(context, obj)
+        context.add_output_metadata({self.env_key: self.staging_label})
```
Comment (sryza) [inline python_modules/dagster/dagster/_experimental/staging_io_manager.py:96]: `add_output_metadata` currently counterintuitively overwrites prior metadata on the context.

Have had a PR out for a while to fix this that I haven't gotten around to merging: https://github.com/dagster-io/dagster/pull/9518.

**Review by sryza — COMMENTED** [review]
Quite clean!

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py (hunk 4 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/staging_io_manager.py:70]: agree

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py (hunk 3 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/staging_io_manager.py:53]: 👍🏻

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py (hunk 5 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/staging_io_manager.py:15]: cool can do in follow up

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py (hunk 6 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/staging_io_manager.py:96]: can you merge or do you want me to take over

Comment (sryza) [reply python_modules/dagster/dagster/_experimental/staging_io_manager.py:72]: I'll do it. Just rebased

File: python_modules/dagster/dagster/_experimental/staging_io_manager.py (hunk 5 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/staging_io_manager.py]: decided to do this in a PR before this one and restack

**Comment by schrockn** [comment]
@sryza This is in a place where we can review again.

Notably changed the name of this to BranchingIOManager, adopted more generic language of "parent" and "branch" rather than "prod" and "staging", and added support for partitions.

File: python_modules/dagster/dagster_tests/storage_tests/branching_io_manager_tests/utils.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    IOManager,
+    InputContext,
+    OutputContext,
+)
+from dagster._core.definitions.events import CoercibleToAssetKey
+from dagster._core.event_api import EventLogRecord, EventRecordsFilter
+from dagster._core.events import DagsterEventType
+
+
+class DefinitionsRunner:
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/storage_tests/branching_io_manager_tests/utils.py:18]: We chatted about this briefly on Zoom, and I don't think it needs to block this PR, but I think this is functionality that many users would find useful, and we should build a public-facing version of it for use in both user tests and notebooks.

Here's a version of this from pre-Definitions world: https://github.com/dagster-io/dagster/pull/10072. To support the case when the `Definitions` has a non-in-process executor, it requires the `Definitions` to be defined at module-scope, but we could relax that constraint when the executor is in-process.

File: python_modules/dagster/dagster/_experimental/branching_io_manager.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+class BranchingIOManager(IOManager):
+    """A branching I/O manager composes two I/O managers.
+
+    1) The upstream I/O manager, typically your production environment.
+    2) The branch I/O manager, typically a development or branched environment.
+
+    The objective of this to allow a developer to safely read from a production
+    environment and then write to a separate development environment, but using
+    the same address space. Once data has been written to the branch environment
+    subsequent reads from that address are sourced from the branch environment. This
```
Comment (sryza) [inline python_modules/dagster/dagster/_experimental/branching_io_manager.py:43]:
```suggestion
    subsequent reads of that asset are sourced from the branch environment. This
```

File: python_modules/dagster/dagster/_experimental/branching_io_manager.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+class BranchingIOManager(IOManager):
+    """A branching I/O manager composes two I/O managers.
+
+    1) The upstream I/O manager, typically your production environment.
+    2) The branch I/O manager, typically a development or branched environment.
+
+    The objective of this to allow a developer to safely read from a production
+    environment and then write to a separate development environment, but using
+    the same address space. Once data has been written to the branch environment
```
[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_experimental/branching_io_manager.py:42]: The meaning of "address space" is a little unclear to me here - I think the sentence might be just as clear without that clause.

File: python_modules/dagster/dagster/_experimental/branching_io_manager.py (hunk 8 above)
Comment (sryza) [inline python_modules/dagster/dagster/_experimental/branching_io_manager.py:43]:
```suggestion
    subsequent reads of that asset are sourced from the branch environment. This
```

Comment (sryza) [inline python_modules/dagster/dagster/_experimental/branching_io_manager.py:43]:
```suggestion
    subsequent reads of that asset are sourced from the branch environment. This
```

**Review by sryza — COMMENTED** [review]
Quite excited about this. I might have gotten a little overzealous on docstring nitpicks.

File: python_modules/dagster/dagster/_experimental/branching_io_manager.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    Some storage systems branching functionality natively. Examples include Snowflake's
+    CLONE feature. Branching I/O managers allow users to implement that functionality
+    in software over arbitrary storage systems."""
+
+    def __init__(
+        self,
+        *,
+        parent_io_manager: IOManager,
+        branch_io_manager: IOManager,
+        branch_name: str = "dev",
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/branching_io_manager.py:60]: Agreed but I'm going to do that in the caller in my demo for now

File: python_modules/dagster/dagster/_experimental/branching_io_manager.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    2) The branch I/O manager, typically a development or branched environment.
+
+    The objective of this to allow a developer to safely read from a production
+    environment and then write to a separate development environment, but using
+    the same address space. Once data has been written to the branch environment
+    subsequent reads from that address are sourced from the branch environment. This
+    bookkeeping is done in Dagster's asset catalog by emitting AssetMaterializations
+    with tags.
+
+    This is extremely useful for iterative development on asset graphs, especially
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/branching_io_manager.py:47]: 👍🏻

File: python_modules/dagster/dagster/_experimental/branching_io_manager.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    where assets early in the graph are large and expensive to compute. One can
+    iteratively develop on downstream assets in that graph safely.
+
+    Some storage systems branching functionality natively. Examples include Snowflake's
+    CLONE feature. Branching I/O managers allow users to implement that functionality
+    in software over arbitrary storage systems."""
```
[the remark this answers is not in the document]
Comment (schrockn) [reply python_modules/dagster/dagster/_experimental/branching_io_manager.py:53]: Went with "Branching I/O managers allow users to implement that functionality in more flexible software layer over arbitrary storage systems."

**Review by sryza — APPROVED** [review]
A place to put this that would be more consistent with our existing experimental IO managers (versioned_filesystem_io_manager, DbIOManager) would be dagster._core.storage.

**Comment by schrockn** [comment]
For postering here's a more fleshed-out example consuming this branch that includes using an environment variable to determine branch name: https://github.com/schrockn/dagster-script-to-assets/blob/working-branch/hn_dagster.py
